index.ts

```diff
@@ -0,0 +1,552 @@
+//   and put their own layer on top which can extend or alter the functionality.
+//
+// CONSIDERATIONS:
+//
+//    - Approach: async APIs are moved to top level so that data load can be done
+//      once and then the rest of the APIs aren't cluttered with `await` everywhere.
+//      Furthermore, individual pieces can be loaded (just `Abilities` vs. all data
+//      types) if desired.
+//    - `getByID` APIs are important for removing the `toID` call in `get` methods
+//      for when we already know through typechecking that we have an ID. String
+//      getters have been removed - callees can manually called `toID` first if they
+//      dont have an ID (which should be very rare unless its human-input).
+//    - Descriptions are UI only and can be large so aren't included in the Dex by
+//      default, but they may be loaded in bulk asynchronously for the desired
+//      languages.
+//    - Similarly, SpeciesDetails which are purely cosmetic or used by the Validator
+//      or for non-competitive reasons are offloaded. eg. learnset details are very
+//      large and not needed in many circumstances.
+//    - All other APIs are loaded together for a gen (+ mod) - it's rare a consumer
+//      cares *just* about Abilities or *just* Moves, and if they do and are in an
+//      environment where loading more is a concern they can implement their own
+//      splitting on top of the raw data files.
+//    - The aliases.json data is used internally (to implement getByID's) but is
+//      not exported.
+//    - The individual APIs (eg. Abilities) are not just Maps, as their lookup function
+//      is more sophisticated (handles aliases, caching, etc).
+//    - Internally, references can be resolved lazily through Javascript getters (though
+//      need to use `Object.defineProperty(Foo.prototype, ...)` to be able to make
+//      enumerable) to avoid object instantiation overhead and avoid creating objects for
+//      fields which are never read.
+//    - Map will need to be polyfilled or support for browsers before IE11 will need to
+//      be dropped.
+//
+//  *** TODO ***:
+//
+//   - 'Species'/'SpeciesDetails' names are problematic as they are plural
+//   - Demonstrate how PS can wrap the specific types and still extend them in OMs
+//   - All the fields in the JSON should be able to be `null` as a sentinel for removal,
+//     can implement this generically using higher order Typescript utility type
+//     functionality.
+//   - 'CommonFoo' and 'JSONFoo' naming could use bikeshedding!
+//   - Review fields to ensure optional fields are made required in the rich APIs where
+//     it makes sense to do so.
+//
+
+// core.ts
+
+type ID = '' | string & {__isID: true};
+toID(s: string): ID;
+
+type Generation = 1 | 2 | 3 | 4 | 5 | 6 | 7;
+// NOTE: CAP is not currently considered a mod but should be...
+type Mod = 'LGPE' | 'Stadium' | 'CAP' | 'SSB' | 'Next' | 'PiC' | 'Mix And Mega' | 'VGC 17';
```

Comment:
Is there something special about VGC 17 wrt other VGC years, or is this just an example?

Reply:
These are the current mods PS has + CAP because I think it belongs as a mod. no fucking clue why VGC 17 needed to be a mod, but https://github.com/Zarel/Pokemon-Showdown/tree/master/data/mods/vgc17 is a thing

Reply:
Seems like its to make some things unreleased/moves unavailable etc?

Reply:
That's correct; mods are used to preserve a specific list of unreleased things.